pkg/storage/ingest/reader_test.go

```diff
@@ -155,40 +155,47 @@ func TestPartitionReader_ConsumerStopping(t *testing.T) {
 
 	_, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
 
-	invocations := atomic.NewInt64(0)
-	blockingConsumer := newTestConsumer(0)
-	consumer := consumerFunc(func(ctx context.Context, records []record) error {
-		invocations.Inc()
-		return blockingConsumer.consume(ctx, records)
+	// consumerErrs will store the last error returned by the consumer; its initial value doesn't matter, but it must be non-nil.
+	consumerErrs := atomic.NewError(errors.New("dummy error"))
+	type consumerCall struct {
+		f    func() []record
+		resp chan error
+	}
+	consumeCalls := make(chan consumerCall)
+	consumer := consumerFunc(func(ctx context.Context, records []record) (err error) {
+		defer consumerErrs.Store(err)
+
+		call := consumerCall{
+			f:    func() []record { return records },
+			resp: make(chan error),
+		}
+		consumeCalls <- call
+		return <-call.resp
 	})
 	reader := createReader(t, clusterAddr, topicName, partitionID, consumer)
 	require.NoError(t, services.StartAndAwaitRunning(ctx, reader))
 
 	// Write to Kafka.
 	writeClient := newKafkaProduceClient(t, clusterAddr)
-
 	produceRecord(ctx, t, writeClient, topicName, partitionID, []byte("1"))
-	produceRecord(ctx, t, writeClient, topicName, partitionID, []byte("2"))
```

Comment:
so my understanding of the deadlock is that record 2 gets produced after record 1 is consumed. So then the reader waits for records 2 to get consumed, which never happens.

Aren't you solving this by just removing record 2? Why are the rest of the changes necessary?

Reply:
Good point about records 2. The idea for the changes was to get rid of any polling for changes in atomic values, and explicitly sync the state of the goroutines instead.

Looking at it now, I don't think this test verifies what we initially wanted. I.e. the idea was to test the behaviour around [this line](https://github.com/grafana/mimir/blob/bcde1f9235c5060580b3c269c3f3393c979a58be/pkg/storage/ingest/reader.go#L418-L426). For that we must synchronize around several key points:

1. the data is already fetched from kafka
2. the service is about to stop
3. the records (and the context) are passed to the consumer

I've updated the test to try make it closer to what we expected to test.